src/data/sonata_data.cpp

```diff
@@ -262,36 +258,50 @@ void SonataData::write_report_header() {
     hdf5_writer_->configure_attribute(reports_population_group + "/data", "units", report_units_);
 
     std::vector<uint64_t> sonata_node_ids(node_ids_);
-    convert_gids_to_sonata(sonata_node_ids);
+    convert_gids_to_sonata(sonata_node_ids, population_offset_);
     hdf5_writer_->write(reports_population_group + "/mapping/node_ids", sonata_node_ids);
     hdf5_writer_->write(reports_population_group + "/mapping/index_pointers", index_pointers_);
     hdf5_writer_->write(reports_population_group + "/mapping/element_ids", element_ids_);
     hdf5_writer_->write_time(reports_population_group + "/mapping/time", time_);
     hdf5_writer_->configure_attribute(reports_population_group + "/mapping/time", "units", "ms");
 }
 
-void SonataData::write_spikes_header(const std::string& order_by) {
+void SonataData::write_spikes_header(const std::unique_ptr<Population>& population) {
     logger->trace("Writing SPIKE header!");
+    const std::string order_by = population->get_sorting();
     if (order_by != "by_time" && order_by != "by_id" && order_by != "none") {
         throw std::runtime_error("Order method " + order_by + "does not exists");
     }
 
-    const std::string spikes_population_group = "/spikes/" + population_name_;
+    const std::string spikes_population_group = "/spikes/" + population->get_population_name();
     hdf5_writer_->configure_group("/spikes");
     hdf5_writer_->configure_group(spikes_population_group);
     hdf5_writer_->configure_enum_attribute(spikes_population_group, "sorting", order_by);
-    hsize_t timestamps_size = Implementation::get_global_dims(report_name_,
-                                                              spike_timestamps_.size());
-    Implementation::sort_spikes(spike_timestamps_, spike_node_ids_, order_by);
-    hdf5_writer_->write(spikes_population_group + "/timestamps", spike_timestamps_);
+    hsize_t timestamps_size =
+        Implementation::get_global_dims(report_name_, population->get_spike_timestamps().size());
+    Implementation::sort_spikes(population->get_spike_timestamps(),
+                                population->get_spike_node_ids(),
+                                order_by);
+    hdf5_writer_->write(spikes_population_group + "/timestamps",
+                        population->get_spike_timestamps());
     if (timestamps_size > 0) {
         hdf5_writer_->configure_attribute(spikes_population_group + "/timestamps", "units", "ms");
     }
-    std::vector<uint64_t> sonata_spike_node_ids(spike_node_ids_);
-    convert_gids_to_sonata(sonata_spike_node_ids);
+    std::vector<uint64_t> sonata_spike_node_ids(population->get_spike_node_ids());
+    convert_gids_to_sonata(sonata_spike_node_ids, population->get_population_offset());
     hdf5_writer_->write(spikes_population_group + "/node_ids", sonata_spike_node_ids);
 }
 
+void SonataData::add_population(std::unique_ptr<Population> population) {
+    populations_.push_back(std::move(population));
```

Comment:
I know there has been a lot of discussion, but this is very unsafe. The C++ way to ask for something we can move is accepting an rvalue reference.
```suggestion
void SonataData::add_population(std::unique_ptr<Population>&& population) {
    populations_.push_back(std::move(population));
```

Reply:
I think it was my suggestion to take `unique_ptr` by value here; I tend to agree with: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-consume
specifically:

> **Exception**
> Unique owner types that are move-only and cheap-to-move, such as unique_ptr, can also be passed by value which is simpler to write and achieves the same effect. Passing by value does generate one extra (cheap) move operation, but prefer simplicity and clarity first.
> For example:
```c++
template<class T>
void sink(std::unique_ptr<T> p)
{
    // use p ... possibly std::move(p) onward somewhere else
}   // p gets destroyed
```

Reply:
To me `std::unique_ptr<Population>&&` makes things clearer on what is expected. The compiler may end up generating the same code anyway :)

Reply:
Also fine, I don't have a strong opinion between `std::unique_ptr<Population>` and `std::unique_ptr<Population>&&`.

Reply:
Done. Changed to ``std::unique_ptr<Population>&&``.